www/app/blog/2025-04-23-drop-column.mdx

```diff
@@ -0,0 +1,264 @@
+      3 | c                            | f            | integer
+(9 rows)
+```
+
+### Manual re-writing
+
+Most of the time, having a column marked as dropped in not a problem at all. But, as you recall from the riddle at the beginning,
+there's a limit to how many columns we can have - and dropped columns count towards that limit.
+
+So, if you need to add a new column, but you can't because of the limit, you will need to manually rewrite the table and get rid of all the dropped columns.
+
+If the table is simple, you can do this with `CREATE TABLE <new_name> AS SELECT ... FROM <old_name>`. Then you can drop the old table and rename the new one.
+
+If the table has constraints, indexes, etc, you can also use `CREATE TABLE <new_name> (LIKE <old_name> INCLUDING ALL)`.
+
+However, even this will not re-create foreign keys and triggers, so you will need to do that manually (in addition to copying the data into the new table).
+
+Finally, you can also use `pg_dump` to dump the table, rename it (in the dump file), and then restore it.
+
+All these methods are both time-consuming and a bit fiddly, so hopefully you'll never drop enough columns to hit the limit.
+
+## Implications for compliance
+
+When I first posted about this on X, I got a lot of comments about how this is a problem for GDPR compliance. I believe the concern is around
+"the right to be forgotten", which is the right to request that personal data be deleted from a company's records.
+
+In my view (I'm not a lawyer, but I have dealt with GDPR compliance before), the right to be forgotten is not relevant to this specific scenario.
+
+If you think about how personal data is stored in Postgres, it is typically stored in various tables, with a foreign key to a user table.
+When a user requests their data or asks to be deleted, the query will look either like this:
+
+```sql
+-- This delete will cascade to all the tables that have a foreign key to the users table
+DELETE FROM users WHERE id = <user_id>;
+```
+
+or like this:
+
+```sql
+-- Delete from the data tables directly
+DELETE FROM data_table_1 WHERE user_id = <user_id>;
+DELETE FROM data_table_2 WHERE user_id = <user_id>;
+-- etc
+DELETE FROM users WHERE id = <user_id>;
+```
+
+In both cases, dropped columns are not the problem, since the user is forgotten by deleting entire rows.
+There are no real-world scenarios where a user "right to be forgotten" request will require dropping columns
+(if you happen to be the exception, `VACUUM FULL` is available, but do consider your data model and whether it actually makes sense).
+
+This raises a different concern: Postgres uses MVCC, so deleted rows are not immediately removed from the table. In fact, they will not be removed until
+all transactions that have seen the row have completed **and** the vacuum process has run.
+Since no reasonable database will leave transactions open and no running vacuum for over a month, this isn't typically a problem.
```

Comment:
there is no special guarantee that vacuum will be triggered withing a month.  If this is a low activity table and you get a single delete, autovacuum is unlikely to be scheduled for that table and unless you run vacuum manually the dead row can stay there for quite a while.  There is also the physical backup considerations.